darwin: Use posix_spawn to spawn subprocesses in macOS

jpcanepa opened the pull request:
# Issue

This PR tackles #3050.

# Overview 

With Big Sur (macOS 11), apple introduced a significant performance degradation when trying to `fork()`/`exec` a new child process when the parent has many pages mmaped with `MAP_JIT` (like, say, a javascript interpreter). At the same time, on macOS spawning a subprocess with `posix_spawn` (as chromium does) does not incur on the overhead. 

This bug impacts electron apps that spawn subprocess, like for example, [VSCode](https://github.com/microsoft/vscode/issues/105446#issuecomment-729274600). There is an issue filed against [electron itself](https://github.com/electron/electron/issues/26143) as a consequence. 

Using the [repro app](https://github.com/deepak1556/libuv-spawn) from the issue, I implemented the forking logic specific to macOS using posix_spawn, as far as I know, doing everything that the normal `fork` path does, with the extra advantage of having the macOS extension to posix_spawn `POSIX_SPAWN_CLOEXEC_DEFAULT` that treats _all_ descriptors, regardless if opened or not with `O_CLOEXEC`, as if they had; that way we can guarantee no descriptors leak into the child process.

pdesantis commented:
I verified that spawn times are just as fast with & without JIT using the repro project (https://github.com/deepak1556/libuv-spawn)!

Results taken using a codesigned build, 2016 MacBook Pro, macOS 11.0.1

# posix_spawn

### Without MAP_JIT

```
bash-3.2$ ./out/Default/uv_spawn
ls process launched with ID : 79108 and in : 1.23733 ms
ls process exited with status : 0 and in : 5.55769 ms

ls process launched with ID : 79113 and in : 1.41789 ms
ls process exited with status : 0 and in : 5.66987 ms

ls process launched with ID : 79118 and in : 0.89105 ms
ls process exited with status : 0 and in : 4.87111 ms

ls process launched with ID : 79123 and in : 0.794623 ms
ls process exited with status : 0 and in : 4.54644 ms

ls process launched with ID : 79128 and in : 0.868151 ms
ls process exited with status : 0 and in : 4.66897 ms

ls process launched with ID : 79133 and in : 0.998443 ms
ls process exited with status : 0 and in : 5.41041 ms

ls process launched with ID : 79138 and in : 1.06477 ms
ls process exited with status : 0 and in : 5.14938 ms

ls process launched with ID : 79143 and in : 0.964673 ms
ls process exited with status : 0 and in : 5.09582 ms

ls process launched with ID : 79148 and in : 0.887238 ms
ls process exited with status : 0 and in : 4.8716 ms

ls process launched with ID : 79153 and in : 0.932572 ms
ls process exited with status : 0 and in : 4.58974 ms
```

### With MAP_JIT

```
bash-3.2$ ./out/Default/uv_spawn --use-jit
ls process launched with ID : 79160 and in : 0.98975 ms
ls process exited with status : 0 and in : 4.46968 ms

ls process launched with ID : 79165 and in : 0.763886 ms
ls process exited with status : 0 and in : 3.7277 ms

ls process launched with ID : 79170 and in : 1.10253 ms
ls process exited with status : 0 and in : 4.79186 ms

ls process launched with ID : 79175 and in : 0.923323 ms
ls process exited with status : 0 and in : 5.28865 ms

ls process launched with ID : 79180 and in : 1.03655 ms
ls process exited with status : 0 and in : 5.19892 ms

ls process launched with ID : 79185 and in : 0.891606 ms
ls process exited with status : 0 and in : 5.07491 ms

ls process launched with ID : 79190 and in : 1.13192 ms
ls process exited with status : 0 and in : 6.28628 ms

ls process launched with ID : 79195 and in : 1.1995 ms
ls process exited with status : 0 and in : 5.56335 ms

ls process launched with ID : 79200 and in : 0.999381 ms
ls process exited with status : 0 and in : 5.27073 ms

ls process launched with ID : 79205 and in : 1.11228 ms
ls process exited with status : 0 and in : 5.3832 ms
```

# fork

Compare the results above against the head of the `v1.x` branch below:

### Without MAP_JIT

```
bash-3.2$ ./out/Default/uv_spawn
ls process launched with ID : 84481 and in : 1.34959 ms
ls process exited with status : 0 and in : 5.45108 ms

ls process launched with ID : 84482 and in : 1.1263 ms
ls process exited with status : 0 and in : 5.045 ms

ls process launched with ID : 84483 and in : 1.02336 ms
ls process exited with status : 0 and in : 5.0336 ms

ls process launched with ID : 84484 and in : 1.09634 ms
ls process exited with status : 0 and in : 4.89075 ms

ls process launched with ID : 84485 and in : 1.00101 ms
ls process exited with status : 0 and in : 4.84648 ms

ls process launched with ID : 84486 and in : 1.80779 ms
ls process exited with status : 0 and in : 6.67846 ms

ls process launched with ID : 84487 and in : 1.26541 ms
ls process exited with status : 0 and in : 5.44206 ms

ls process launched with ID : 84488 and in : 1.02727 ms
ls process exited with status : 0 and in : 4.8705 ms

ls process launched with ID : 84489 and in : 1.07131 ms
ls process exited with status : 0 and in : 5.43011 ms

ls process launched with ID : 84490 and in : 1.07475 ms
ls process exited with status : 0 and in : 4.72903 ms
```

### With MAP_JIT

```
bash-3.2$ ./out/Default/uv_spawn --use-jit
ls process launched with ID : 84494 and in : 53.6429 ms
ls process exited with status : 0 and in : 57.4913 ms

ls process launched with ID : 84495 and in : 45.3467 ms
ls process exited with status : 0 and in : 48.786 ms

ls process launched with ID : 84496 and in : 44.0292 ms
ls process exited with status : 0 and in : 47.2501 ms

ls process launched with ID : 84497 and in : 45.9806 ms
ls process exited with status : 0 and in : 49.2605 ms

ls process launched with ID : 84498 and in : 46.071 ms
ls process exited with status : 0 and in : 49.3887 ms

ls process launched with ID : 84499 and in : 43.6015 ms
ls process exited with status : 0 and in : 47.0534 ms

ls process launched with ID : 84500 and in : 47.0892 ms
ls process exited with status : 0 and in : 50.5227 ms

ls process launched with ID : 84501 and in : 55.4421 ms
ls process exited with status : 0 and in : 59.0695 ms

ls process launched with ID : 84502 and in : 45.3965 ms
ls process exited with status : 0 and in : 48.9634 ms

ls process launched with ID : 84503 and in : 49.845 ms
ls process exited with status : 0 and in : 53.2581 ms
```

jpcanepa commented:
Switched to a proper PR. Updated the description to include all the information I have about the successful/failed builds and their respective command lines emited by the build systems.

jpcanepa commented:
Ok, with the help of @marcello3d the build is now back up, and so I could run the unit tests and fix all but one. I've rewritten the PR description to reflect the current state of affairs.

bnoordhuis commented on src/unix/darwin-stub.h:
```diff
@@ -110,4 +111,9 @@ static const int kFSEventStreamEventFlagRootChanged = 32;
 static const int kFSEventStreamEventFlagUnmount = 128;
 static const int kFSEventStreamEventFlagUserDropped = 2;
 
+// Copied from https://opensource.apple.com/source/xnu/xnu-6153.101.6/libsyscall/wrappers/spawn/spawn_private.h.auto.html
```
```suggestion
/* Copied from https://opensource.apple.com/source/xnu/xnu-6153.101.6/libsyscall/wrappers/spawn/spawn_private.h.auto.html */
```
I.e., no C99/C++ style comments. Also applies to other files.

Also, as a general rule, wrap lines > 80 columns.

bnoordhuis commented on src/unix/darwin-stub.h:
```diff
@@ -110,4 +111,9 @@ static const int kFSEventStreamEventFlagRootChanged = 32;
 static const int kFSEventStreamEventFlagUserDropped = 2;
 
+// Copied from https://opensource.apple.com/source/xnu/xnu-6153.101.6/libsyscall/wrappers/spawn/spawn_private.h.auto.html
+int     posix_spawnattr_set_uid_np(const posix_spawnattr_t * __restrict, uid_t) __API_AVAILABLE(macos(10.15), ios(13.0), tvos(13.0), watchos(6.0));
+int     posix_spawnattr_set_gid_np(const posix_spawnattr_t * __restrict, gid_t) __API_AVAILABLE(macos(10.15), ios(13.0), tvos(13.0), watchos(6.0));
+int     posix_spawnattr_set_groups_np(const posix_spawnattr_t * __restrict, int, gid_t * __restrict, uid_t) __API_AVAILABLE(macos(10.15), ios(13.0), tvos(13.0), watchos(6.0));
```
```suggestion
int posix_spawnattr_set_uid_np(const posix_spawnattr_t*, uid_t);
int posix_spawnattr_set_gid_np(const posix_spawnattr_t*, gid_t);
int posix_spawnattr_set_groups_np(const posix_spawnattr_t*, int, gid_t*, uid_t);
```

pdesantis replied:
@bnoordhuis FYI after removing the `__API_AVAILABLE` macro, I get the following error when compiling this branch of libuv with Electron:

```
../../third_party/electron_node/deps/uv/src/unix/process.c:411:31: error: 'posix_spawn_file_actions_addchdir_np' is only available on macOS 10.15 or newer [-Werror,-Wunguarded-availability-new]
  if (options->cwd != NULL && posix_spawn_file_actions_addchdir_np(&actions, options->cwd))
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../../../../../.electron_build_tools/third_party/Xcode/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/usr/include/spawn.h:178:9: note: 'posix_spawn_file_actions_addchdir_np' has been marked as being introduced in macOS 10.15 here, but the deployment target is macOS 10.10.0
int     posix_spawn_file_actions_addchdir_np(posix_spawn_file_actions_t *,
        ^
../../third_party/electron_node/deps/uv/src/unix/process.c:411:31: note: enclose 'posix_spawn_file_actions_addchdir_np' in a __builtin_available check to silence this warning
  if (options->cwd != NULL && posix_spawn_file_actions_addchdir_np(&actions, options->cwd))
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Is there a preferred way in libuv of specifying the API availability? If not, no worries, it's easy enough to get Electron to compile w/ libuv by adding the `__API_AVAILABLE` macro back in as a patch in Electron.

bnoordhuis replied:
> Is there a preferred way in libuv of specifying the API availability?

In fact, yes :-) - `dlsym(RTLD_DEFAULT, "posix_spawn_file_actions_addchdir_np")`

Libuv generally uses `dlsym()` for things that might or might not be available at runtime. We try to avoid compile-time feature checks because of binary compatibility - a binary compiled on a newer system should, if at all possible, Just Work on an older system.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -424,20 +426,162 @@ int uv_spawn(uv_loop_t* loop,
+    posix_spawn_file_actions_t actions;
+
+    err = posix_spawnattr_init(&attrs);
+    if(err)
```
```suggestion
    if (err != 0)
```
(I won't point out more individual instances but please fix them all.)

bnoordhuis commented on src/unix/process.c:
```diff
@@ -424,20 +426,162 @@ int uv_spawn(uv_loop_t* loop,
-  if (pid == 0) {
-    uv__process_child_init(options, stdio_count, pipes, signal_pipe[1]);
-    abort();
+    if ((options->flags & UV_PROCESS_SETUID) && posix_spawnattr_set_uid_np(&attrs, options->uid)) {
+      goto error;
+    }
```
More idiomatic libuv style is this:
```suggestion
    if (options->flags & UV_PROCESS_SETUID)
      if (posix_spawnattr_set_uid_np(&attrs, options->uid))
        goto error;
```

bnoordhuis commented on src/unix/process.c:
```diff
@@ -424,20 +426,162 @@ int uv_spawn(uv_loop_t* loop,
+     * will be reset to behave as their default
+     * 3) POSIX_SPAWN_SETSIGMASK: Signal mask will be set to the value of spawn-sigmask in attributes
+     * 4) POSIX_SPAWN_SETSID: Make the process a new session leader if a detached session was requested. */
+    err = posix_spawnattr_setflags(&attrs, 
+      POSIX_SPAWN_CLOEXEC_DEFAULT |
+      POSIX_SPAWN_SETSIGDEF |
+      POSIX_SPAWN_SETSIGMASK |
+      (options->flags & UV_PROCESS_DETACHED ? POSIX_SPAWN_SETSID : 0)
+      );
```
Easier to read when written like this:
```suggestion
    flags = POSIX_SPAWN_CLOEXEC_DEFAULT |
      POSIX_SPAWN_SETSIGDEF |
      POSIX_SPAWN_SETSIGMASK;

    if (options->flags & UV_PROCESS_DETACHED)
      flags |= POSIX_SPAWN_SETSID;

     if (posix_spawnattr_setflags(&attrs, flags))
       goto error;
```

bnoordhuis reviewed (commented):
I did a very light review, mostly on style issues.

I feel the big `#ifdef __APPLE__` block in the middle of an already large function is not so nice, it makes things hard to follow.

It should probably be broken out into a separate function, possibly in a different file like `src/unix/darwin.c`.

jpcanepa commented:
Moved the process spawn and initialization logic to a separate function which in turn delegates to a purpose specific one (posix_spawn or fork+exec). Also, corrected all formatting and convenction issues.

jpcanepa commented:
Uh oh. Apparently CI did not like the changes to the stubs.

jpcanepa commented:
Thanks @bnoordhuis for the review! I've addressed all comments.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -342,6 +345,181 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  unsigned int flags;
+
+  err = posix_spawnattr_init(&attrs);
+  if(err != 0)
```
Stylistic: `if (err != 0)` - note the space. Applies globally.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -342,6 +345,181 @@ static void uv__process_child_init(const uv_process_options_t* options,
+    * 2) POSIX_SPAWN_SETSIGDEF: Signals mentioned in spawn-sigdefault in the spawn attributes
+    * will be reset to behave as their default
+    * 3) POSIX_SPAWN_SETSIGMASK: Signal mask will be set to the value of spawn-sigmask in attributes
+    * 4) POSIX_SPAWN_SETSID: Make the process a new session leader if a detached session was requested. */
```
Long lines.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -342,6 +345,181 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  if(err != 0)
+    return err;
+
+  // Reset all signal the child to their default behavior
```
C89 comments only, please. Applies here and elsewhere.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -342,6 +345,181 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  if(err != 0)
+    return err;
+
+  if (options->cwd != NULL && posix_spawn_file_actions_addchdir_np(&actions, options->cwd))
```
Long line. Use the nested-if paradigm.

bnoordhuis commented on src/unix/process.c:
```diff
@@ -342,6 +345,181 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  } 
+
+  // Preserve parent environment if not explicitly set
+  char ** env = options->env ? options->env : environ;
```
```suggestion
  char** env = options->env ? options->env : environ;
```

bnoordhuis commented on src/unix/process.c:
```diff
@@ -342,6 +345,181 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  // Spawn the child
+  err = posix_spawnp(pid, options->file, &actions, &attrs, options->args, env);
+
+  return UV__ERR(err);
```
Resource leak: neither the happy path nor the error paths call `posix_spawnattr_destroy()`.

bnoordhuis reviewed (commented):
I don't expect to have much time in the coming weeks so it would be good if other maintainers can also help review.

jpcanepa commented:
Resolved all pointed out issues. I refactored the initialization of the `posix_spawnattr` and `posix_spawn_file_actions` to separate functions that either return a properly initialized object on success, or an already-destroyed one on failure, making the cleanup in the top level more simple and also the main happy path more readable.

jpcanepa commented on src/unix/process.c:
```diff
@@ -342,6 +345,239 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  err = posix_spawnp(pid, options->file, &actions, &attrs, options->args, env);
+
+  /* Destroy the actions/attributes */
+  (void) posix_spawn_file_actions_destroy(&actions);
+  (void) posix_spawnattr_destroy(&attrs);
```
Seemed to me that the failure of either of these functions was not a reason to signal failure to the caller.

vtjnash replied:
optional: I'd perhaps suggest making them call abort instead?

jpcanepa replied:
Your call, @vtjnash. If you think that failure of these calls warrants termination, let's do that.

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+                                                pipes,
+                                                pid,
+                                                &posix_spawn_fncs);
+  } else {
+#endif  
+    return uv__spawn_and_init_child_fork(options, stdio_count, pipes, error_fd, pid);
+#if defined(__APPLE__) 
+  }
+#endif
```
style nit:
```suggestion
  }
  #endif  
  return uv__spawn_and_init_child_fork(options, stdio_count, pipes, error_fd, pid);
```

vtjnash commented on src/unix/process.c:
```diff
@@ -34,8 +34,12 @@
 #include <poll.h>
 
 #if defined(__APPLE__) && !TARGET_OS_IPHONE
+#include <spawn.h>
+#include <sys/kauth.h>
+#include <dlfcn.h>
 # include <crt_externs.h>
 # define environ (*_NSGetEnviron())
+#include "darwin-stub.h"
```
indentation fix:
```suggestion
# include <sys/kauth.h>
# include <dlfcn.h>
# include <crt_externs.h>
# define environ (*_NSGetEnviron())
# include "darwin-stub.h"
```
why's spawn.h listed in both files? actually, seems possible it's only needed here?

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  fncs->spawnattr.set_groups_np = dlsym(RTLD_DEFAULT, "posix_spawnattr_set_groups_np");
+  fncs->file_actions.addchdir_np = dlsym(RTLD_DEFAULT, "posix_spawn_file_actions_addchdir_np");
+
+  /* If any of these is missing, just don't use posix_spawn */
```
These are all fairly new (10.15), which means we currently can't test this code on libuv CI (or use it on other platforms). Could we instead selectively enable the posix_spawn code for the common cases were we don't need to use any of these? (return ENOSYS from uv__spawn_and_init_child_posix_spawn, then have that bubble up to uv__spawn_and_init_child where it can use the fork-exec next strategy)

jpcanepa replied:
I've implemented this alternative.

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  if (options->flags & (UV_PROCESS_SETUID | UV_PROCESS_SETGID)) {
+    /* See the comment on the call to setgroups in uv__process_child_init above
+      * for why this is not a fatal error */
+    SAVE_ERRNO(posix_spawn_fncs->spawnattr.set_groups_np(attrs, 0, NULL, KAUTH_UID_NONE));
```
Is there any documentation on this? The best I could find is the source code, which indicates that NULL is invalid:

https://github.com/Hardercoder/opensourceapple-tarballs/blob/235bdeaf60d803fb88de0da8d5a53d67e7eaafd0/xnu-6153.11.26/libsyscall/wrappers/spawn/posix_spawn.c#L2193

jpcanepa replied:
None that I could find, hence the caveat in the description. I used the source for reference too, but failed to check that restriction. The ngroups set to zero means that that var will never be read (besides the if call, of course), so I think we can just pass in a pointer to a gid_t and that should be enough...

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+  /* First, dupe any required fd into orbit, out of the range of 
+   * the descriptors that should be mapped in. */
+  for(fd = 0 ; fd < stdio_count; ++fd) {
```
style nit:
```suggestion
  for (fd = 0; fd < stdio_count; fd++) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+      goto error;
+  }
+
+  /* First, dupe any required fd into orbit, out of the range of 
```
```suggestion
  /* First, duplicate any required fd into orbit, out of the range of 
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  }
+
+  /*  Second, move the descriptors into their respective places */
+  for(fd = 0 ; fd < stdio_count; ++fd) {
```
style nit
```suggestion
  for (fd = 0; fd < stdio_count; fd++) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  }
+
+  /*  Finally, close all the superfluous descriptors */
+  for(fd = 0; fd < stdio_count; ++fd) {
```
style nit
```suggestion
  for (fd = 0; fd < stdio_count; fd++) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  }
+
+  /*  Finally process the standard streams as per de documentation */
+  for(fd = 0 ; fd < 3 ; ++fd) {
```
style nit:
```suggestion
  for (fd = 0; fd < 3; fd++) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+      goto error;
+  }
+
+  /*  Finally process the standard streams as per de documentation */
```
```suggestion
  /*  Finally process the standard streams as per documentation */
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /*  Finally process the standard streams as per de documentation */
+  for(fd = 0 ; fd < 3 ; ++fd) {
+    /*  If ignored, open as /dev/null */
+    const int oflags = fd == 0 ? O_RDONLY : O_RDWR;
+    const int mode = 0;
```
separate code and variables:
```suggestion
    int oflags;
    const int mode = 0;
    oflags = fd == 0 ? O_RDONLY : O_RDWR;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+      goto error;
+  }  
+
+  return err;
```
```suggestion
  return 0;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+                             pid_t* pid) {
+
+#if defined(__APPLE__) 
+  uv__posix_spawn_fncs_t posix_spawn_fncs;
+  if (uv__spawn_use_posix_spawn(&posix_spawn_fncs)) {
```
This can be made a global, initialized with `uv_once`

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
+     * logic with poxis_spawn and its related primitves. It also takes advantage of 
+     * the macOS extension POSIX_SPAWN_CLOEXEC_DEFAULT that makes impossible to
+     * leak descriptors to the child process.
+     * 
+     * see https://github.com/libuv/libuv/issues/3050
```
issue links are in git blame, so aren't included in comments

vtjnash commented:
Reviewed this, and looks nearly done. Note it also fixes https://github.com/joyent/libuv/issues/1483

> There is one unit test failure: spawn_reads_child_path
You might want to look at the `execvpe` emulation in the file?

vtjnash commented:
I realized the code for `execvpe` emulation (support for older glibc) isn't actually available in this fork of libuv, so that comment wasn't helpful. Here's the code I intended to reference: https://github.com/JuliaLang/libuv/commit/9af3af617138d6a6de7d72819ed362996ff255d9

andreialecu commented:
Sorry for the bump, but I was wondering - because it now seems complete - if there were any updates about landing this?

Visual Studio Code on Apple Silicon (which only works on Big Sur), is pretty degraded because of this issue. 😕 

Thanks!

jpcanepa commented:
@vtjnash I though doing something like that, but I was concerned with potential issues with path handling and unicode paths, but if it is a real alternative (I've been looking at the source of execve from glibc, and it is essentially the same. So I'll go ahead an mimic that behavior.

jpcanepa commented:
@vtjnash I've been going over the emulation of the execvpe path resolution, but I'm not sure that we can replicate all the possible scenarios. 

We have no control over the environment that posix_spawn uses to resolve the file to be executed (we can't manipulate `environ` as that would break concurrent operations): it will just use the environment of the current image. 

Most cases I think will be ok, but there is one case that I'm not sure how to handle: given how the fork() flow works, the search that `execvp` does is over the provided environment _only_ if any is supplied (since it replaces `environ` with `options->env` in the child init func). So if the caller supplied an environment with _no_ `PATH` env, execve would have no PATH to resolve the file if it does not contain an `/`. 

In the same scenario, but in the `posix_spawn` flow, I cannot mimic that particular case. I can try to resolve the file in the supplied env `PATH` if any, but if none is provided, I cannot emulate `posix_spawnp` trying to resolve the file in an empty path, as it will always pick up the `PATH` from the parent process. 

Yes, it would be a much subtle difference and probably a corner case, but different nevertheless. 

I'll continue working on this emulation, but would love some feedback on this potential limitation.

vtjnash commented:
If no PATH exists, then the default path (`_PATH_DEFPATH`) should be used in the emulation.

The glibc version should not be referenced as it is GPL3. The code I linked above is MIT-licensed (from musl) and can be included directly. I believe it already handles the cases you're concerned about, just replacing `execve` with `posix_spawn`. This is only needed when the user passes options->env. Otherwise, it can use `posix_spawnp`.

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +345,303 @@ static void uv__process_child_init(const uv_process_options_t* options,
+                                           const uv__posix_spawn_fncs_t* posix_spawn_fncs,
+                                           const uv_process_options_t* options,
+                                           int stdio_count,
+                                           int (*pipes)[2])
+{
```
style nit:
```suggestion
                                           int (*pipes)[2]) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +345,303 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  (void) posix_spawnattr_destroy(attrs);
+  return err;
+}
+
```
style nit: two spaces between functions
```suggestion


```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +345,303 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+int uv__spawn_set_posix_spawn_attrs(posix_spawnattr_t* attrs,
+                                    const uv__posix_spawn_fncs_t* posix_spawn_fncs,
+                                    const uv_process_options_t* options)
+{
```
style nit
```suggestion
                                    const uv_process_options_t* options) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +345,303 @@ static void uv__process_child_init(const uv_process_options_t* options,
+   * 
+   * On macOS, though, posix_spawn is implemented in a way that does not 
+   * exhibit the problem. This block implements the forking and preparation
+   * logic with poxis_spawn and its related primitves. It also takes advantage of 
```
```suggestion
   * logic with posix_spawn and its related primitives. It also takes advantage of 
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +345,303 @@ static void uv__process_child_init(const uv_process_options_t* options,
+   * exhibit the problem. This block implements the forking and preparation
+   * logic with poxis_spawn and its related primitves. It also takes advantage of 
+   * the macOS extension POSIX_SPAWN_CLOEXEC_DEFAULT that makes impossible to
+   * leak descriptors to the child process.
+   */
```
```suggestion
   * leak descriptors to the child process. */
```

vtjnash commented on src/unix/process.c:
```diff
@@ -342,6 +345,303 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /* The posix_spawn flow will return UV_ENOSYS if any of the posix_spawn_x_np
+   * non-standard functions is both _needed_ and _undefined_. In those cases, 
+   * default back to the fork/execve strategy. For all other errors, just fail. */
+  if(err != UV_ENOSYS) {
```
```suggestion
  if (err != UV_ENOSYS) {
```

vtjnash reviewed (commented):
Just the reads-child-PATH and a few small remaining style nits, and I think this will be ready.

jpcanepa commented:
@vtjnash This last set of commits fix all the style issues and also make the unit test pass.

🚨 Still, we are investigating with @marcello3d and @pdesantis a slowdown when using this patched version of libuv in node that manifests in poor throughput when piping data between std streams (stdout -> stdin) of spawned processes. It is reproducible when consuming in node when using this branch of libuv, but have not (yet) been able to reproduce with libuv by itself. So please, don't merge just yet.

marcello3d commented:
I don't have a repro for the bug in libuv itself, but I can repro with a custom node.js build ([nodejs branch/changes](https://github.com/nodejs/node/compare/v12.18.3...descriptinc:mbf/libuv-posix-spawn)).

I get a 30x slowdown with the following javascript (requires `ffmpeg` installed):

```js
const { spawn } = require('child_process');
const targetProc = spawn('wc');
spawn(
    'ffmpeg',
    ['-f', 'lavfi', '-i', 'testsrc=duration=2:size=1920x1080:rate=30', '-f', 'rawvideo', '-'],
    { stdio: ['pipe', targetProc.stdin, process.stderr] },
).on('exit', () => {
    console.log('decoder exited');
    targetProc.stdin.end();
});
```

vanilla node 12.18.3 (to match the Electron version we're using):
```
frame=   60 fps= 36 q=-0.0 Lsize=  364500kB time=00:00:02.00 bitrate=1492992.0kbits/s speed= 1.2x    
video:364500kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%
decoder exited

real	0m1.818s
user	0m1.583s
sys	0m0.406s
```

with patches:
```
frame=   60 fps=1.1 q=-0.0 Lsize=  364500kB time=00:00:02.00 bitrate=1492992.0kbits/s speed=0.0367x    
video:364500kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%
decoder exited

real	0m54.616s
user	0m2.026s
sys	0m1.058s
```

marcello3d commented:
Does anyone have an idea why this might happen? It seems like something about how the pipes are being configured with posix spawn is just wrong…

Here's another strange test case I found that doesn't exhibit the performance problem, but seems to terminate the pipe early:
```js
const { spawn } = require('child_process');
const destination = spawn('wc', ['-l']);
destination.stdout.pipe(process.stdout);

// Just need a large file passed in to repro, using /bin/bash binary as a test
spawn('cat', ['/bin/bash'], {
    stdio: ['pipe', destination.stdin, process.stderr],
}).on('exit', () => {
    destination.stdin.end();
});
```

with vanilla node:
```
 $ time node test-pipe.js 
    1858

real	0m0.127s
user	0m0.062s
sys	0m0.059s
```

with the patched node it seems to crash with this interesting error "cat: stdout: Resource temporarily unavailable":
```
[LIBUV] USING POSIX_SPAWN (v2)
[LIBUV] USING POSIX_SPAWN (v2)
cat: stdout: Resource temporarily unavailable
      12

real	0m0.048s
user	0m0.034s
sys	0m0.013s
```

vtjnash commented:
That is usually EAGAIN, which would mean a pipe was incorrectly put in non-blocking mode.

jpcanepa commented:
Ok, I have something. I reproduced the slowdown in the fork flow too by commenting out [this line](https://github.com/libuv/libuv/blob/493be3ed2347cbb61651c878620360d5174348b9/src/unix/process.c#L266), which means very likely right, @vtjnash, with respect to the non-blocking mode. 

Now my problem is that as far as I know there are no posix_spawn file actions fo invoke random fnctls or to mark the fd as non blocking, so I don't know how to replicate that behavior on this alternate flow.

vtjnash commented:
Ahah then, that seems like this known issue: https://github.com/libuv/libuv/issues/2062. Since that flag is on the file description, not the fd, you should be able to call that in the parent before calling posix_spawn and the result will be exactly the same.

jpcanepa commented:
Awesome. Calling uv__nonblock_fcntl on the descriptors resolved the problem in all our tests. I think we are done here.

pdesantis commented:
Thanks for the help @vtjnash! JP implemented that change (https://github.com/libuv/libuv/pull/3064/commits/9a3d751d67807cfa1377c9f9bff35208f8402f0c), and he & Marcello verified it fixed the slow pipes in our test scripts listed above. I'm rebuilding Electron now to verify it there.

JP is looking in to writing a unit test to verify this behavior. However, as far as we know this doesn't need any more functional changes!

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +347,404 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /* The environment for the child process is that of the parent unless overriden 
+   * by options->env */
+  char** env = environ;
+  if (options->env != NULL) {
+    env = options->env;
+  }
```
Style nit:
```suggestion
  if (options->env != NULL)
    env = options->env;
```

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +347,404 @@ static void uv__process_child_init(const uv_process_options_t* options,
+   * to be used, the environment used for path  resolution as well for the 
+   * child process is that of the parent process, so posix_spawnp is the
+   * way to go. */
+  if (strchr(options->file, '/') != NULL || options->env == NULL) {
+    return posix_spawnp(pid, options->file, actions, attrs, options->args, env);
+  }
```
Style nit:
```suggestion
  if (strchr(options->file, '/') != NULL || options->env == NULL)
    return posix_spawnp(pid, options->file, actions, attrs, options->args, env);
```

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +347,404 @@ static void uv__process_child_init(const uv_process_options_t* options,
+   * 
+   * On macOS, though, posix_spawn is implemented in a way that does not 
+   * exhibit the problem. This block implements the forking and preparation
+   * logic with posix_spawn and its related primitves. It also takes advantage of
```
```suggestion
   * logic with posix_spawn and its related primitives. It also takes advantage of
```

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +347,404 @@ static void uv__process_child_init(const uv_process_options_t* options,
+#if defined(__APPLE__) 
+  uv_once(&posix_spawn_init_fncs_once, uv__spawn_init_posix_spawn_fncs);
+
+  /* Especial child process spawn case for macOS Big Sur (11.0) onwards 
```
```suggestion
  /* Special child process spawn case for macOS Big Sur (11.0) onwards 
```

santigimeno reviewed (changes requested):
Left some style related comments.

jpcanepa commented:
@santigimeno thanks for the review. I think I addressed all the style nits.

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+                                posix_spawnattr_t* attrs,
+                                posix_spawn_file_actions_t* actions,
+                                pid_t* pid) {
+  const char *p, *z, *path = NULL;
+  size_t l, k;
```
style nit:
```suggestion
  const char *p;
  const char *z;
  const char *path = NULL;
  size_t l;
  size_t k;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /* If options->file contains a slash, posix_spawn/posix_spawnp behave
+   * the same, and don't involve PATH resolution at all. Otherwise, if
+   * options->file does not include a slash, but no custom environment is 
+   * to be used, the environment used for path  resolution as well for the 
```
```suggestion
   * to be used, the environment used for path resolution as well for the 
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+  l = strnlen(path, PATH_MAX-1)+1;
+
+  for(p = path;; p = z) {
```
```suggestion
  for (p = path;; p = z) {
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+  k = strnlen(options->file, NAME_MAX+1);
+  if (k > NAME_MAX) {
+    errno = ENAMETOOLONG;
+    return -1;
```
```suggestion
    return ENAMETOOLONG;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+  /* Short circuit for erroneous case */
+  if (options->file == NULL) 
+    return UV_ENOENT;
```
```suggestion
    return ENOENT;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /* Set the current working directory if requested */
+  if (options->cwd != NULL) {
+    if (posix_spawn_fncs->file_actions.addchdir_np == NULL) {
+      err = UV_ENOSYS;
```
```suggestion
      err = ENOSYS;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+    gid_t group_array = KAUTH_GID_NONE;
+
+    if (posix_spawn_fncs->spawnattr.set_groups_np == NULL) {
+      err = UV_ENOSYS;
```
```suggestion
      err = ENOSYS;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+  if (options->flags & UV_PROCESS_SETGID) {
+    if (posix_spawn_fncs->spawnattr.set_gid_np == NULL) {
+      err = UV_ENOSYS;
```
```suggestion
      err = ENOSYS;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+  if (options->flags & UV_PROCESS_SETUID) {
+    if (posix_spawn_fncs->spawnattr.set_uid_np == NULL) {
+      err = UV_ENOSYS;
```
```suggestion
      err = ENOSYS;
```

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  for(p = path;; p = z) {
+    /* Compose the new process file from the entry in the PATH
+     * environment variable and the actual file name */
+    char b[l+k+1];
```
VLAs are a C99 feature, and not always well supported by compilers today. I think this should be PATH_MAX or malloc?

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,424 @@ static void uv__process_child_init(const uv_process_options_t* options,
+    err = posix_spawn_file_actions_adddup2(
+      actions, 
+      pipes[fd][1], 
+      stdio_count + fd
+    );
```
style nit:
```suggestion
      stdio_count + fd);
```

vtjnash reviewed (commented):
Looks like there might be a few incorrect constants being returned, and few minor style issues remaining, but I expect it is close.

jpcanepa commented:
Thanks again, @vtjnash. I think I addressed them all. I had to rewrite a little the path routine because the new one had some issues, but I kept the spirit of the suggestion.

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,423 @@ static void uv__process_child_init(const uv_process_options_t* options,
+   * provided env array, and return its value if found */
+  for (env_iterator = env; *env_iterator != NULL; env_iterator++) {
+    const char path_var[] = "PATH=";
+    if(strncmp(*env_iterator, path_var, sizeof(path_var) - 1) == 0) {
```
```suggestion
    if (strncmp(*env_iterator, path_var, sizeof(path_var) - 1) == 0) {
```

vtjnash reviewed (approved):
CI: https://ci.nodejs.org/job/libuv-test-commit/2111/

vtjnash commented on src/unix/process.c:
```diff
@@ -343,6 +346,423 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  flags = POSIX_SPAWN_CLOEXEC_DEFAULT |
+          POSIX_SPAWN_SETSIGDEF |
+          POSIX_SPAWN_SETSIGMASK;
+  if (options->flags & UV_PROCESS_DETACHED) 
+    flags |= POSIX_SPAWN_SETSID;
```
This appears to have been unavailable until 10.15, and so will need to be defined manually and checked for at runtime

jpcanepa replied:
Could you advise as what you expect in runtime? Conditionally executing with `__builtin_available` is not allowed, as far as I understand.  At _compile_ time, I could wrap this logic in `#ifndef POSIX_SPAWN_SETSID` and have the else side return `ENOSYS` to fall back to the fork flow, resolving builds done on 10.14. But I'm not sure how to runtime-check this one.

bnoordhuis replied:
This one is problematic. You can use the constant literal value (i.e. `1024 /* POSIX_SPAWN_SETSID */`) but the xnu kernel doesn't report errors for flags it doesn't recognize - posix_spawn silently spawns the process without calling setsid.

There's a mach system call you can use to find out the kernel version. It's the same output as `uname -v`.
```c
#include <mach/mach_host.h>
#include <stdio.h>
int main(void) {
  kernel_version_t version = {0};
  host_kernel_version(mach_host_self(), version);
  puts(version);
}
```
On my system prints:
```
Darwin Kernel Version 18.7.0: Thu Jun 20 18:42:21 PDT 2019; root:xnu-4903.270.47~4/RELEASE_X86_64
```

vtjnash replied:
You can check that with `sysctlbyname("kern.osrelease", str, &size, NULL, 0);` >= `19.0.0`

https://stackoverflow.com/questions/11072804/how-do-i-determine-the-os-version-at-runtime-in-os-x-or-ios-without-using-gesta

jpcanepa replied:
I have implemented the `sysctlbyname` thing, and it seems to work just fine. We are running some tests on 10.14 after we though we had ran into some runtime issues, but I think I'll push the change very soon.

jpcanepa replied:
Ok, pushed.

jpcanepa commented:
@vtjnash @bnoordhuis @santigimeno Ok, I think this is it.

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +354,474 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+static uv_once_t posix_spawn_init_once = UV_ONCE_INIT;
+static uv__posix_spawn_fncs_t posix_spawn_fncs;
+static int posix_spawn_can_use_setsid = 0;
```
No need to initialize an `static` variable to 0:
```suggestion
static int posix_spawn_can_use_setsid;
```

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +354,474 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  /* Look for an environment variable called PATH in the 
+   * provided env array, and return its value if found */
+  for (env_iterator = env; *env_iterator != NULL; env_iterator++) {
+    const char path_var[] = "PATH=";
+    if (strncmp(*env_iterator, path_var, sizeof(path_var) - 1) == 0) {
+      /* Found "PATH=" at the beginning of the string */
+      return *env_iterator + sizeof(path_var) - 1;
```
I guess the compiler should optimize, but what about moving the `char_path[]` declaration outside the loop?

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +354,474 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  const char *path = NULL;
+  size_t l;
+  size_t k;
+  int err = -1;
```
Style nit: separate declaration from assignment

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +354,474 @@ static void uv__process_child_init(const uv_process_options_t* options,
+
+  /* If no path was provided in options->env, use the default value 
+   * to look for the executable */
+  if (!path) 
```
```suggestion
  if (path == NULL) 
```

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +354,474 @@ static void uv__process_child_init(const uv_process_options_t* options,
+                             int (*pipes)[2],
+                             int error_fd,
+                             pid_t* pid) {
+  int err = 0;
```
```suggestion
  int err;
```

santigimeno commented on src/unix/process.c:
```diff
@@ -343,6 +354,474 @@ static void uv__process_child_init(const uv_process_options_t* options,
+  int r;
+  int version_major;
+
+  /* By default, assume failure */
+  posix_spawn_can_use_setsid = 0;
```
Is this needed? Shouldn't it be always `0` ?

jpcanepa replied:
Not initializing variable makes me nervous, and (to me at least) having the default case value right there makes it easier to follow the code.

santigimeno replied:
I'm fine either way. As I said, they were non-blocking comments

santigimeno reviewed (changes requested):
Cannot test at this moment but looks good. Left some non-blocking comments

jpcanepa commented:
@santigimeno Thanks, addressed all but one, with had an open question to which I replied; but I'll change if requested.

fenvalle commented:
is it ready to merge?!

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2117/

jpcanepa commented:
@santigimeno The output of Jenkins shows failures, but it seems it is mostly unit tests that are unrelated to this change. Or is there anything that needs to be done here too?

pdesantis commented:
@vtjnash @bnoordhuis @santigimeno a friendly heads up - we'll be on vacation for the first 3 weeks of February. So if possible, let us know soon if there are any more changes required, otherwise we won't be able to implement them until later next month. Thanks again for the help!

andrejgloot commented:
Any chance @santigimeno and @bnoordhuis could have a look?

santigimeno reviewed (approved):
LGTM. One last CI run: https://ci.nodejs.org/job/libuv-test-commit-linux/2176/

santigimeno commented:
@jpcanepa could you rebase to current `v1.x`  and squash the commits with a proper commit message? Thanks!!

pdesantis commented:
@santigimeno done!

cjihrig commented:
FYI - we had to revert this because it was causing failures in the Node.js test suite.

KishanBagaria commented:
> FYI - we had to revert this because it was causing failures in the Node.js test suite.

Can you link to those test failures? Figuring out if we can fork and apply the patch without issues.
